File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
         int mx=0;
    unordered_map<char,int>map;
    int st=0;
    for (int i = 0; i <s.length() ; ++i) {
        if(map.find(s[i])!=map.end()){
            if(map[s[i]]>=st){
                st=map[s[i]]+1;
                mx = max(mx,i-st+1);
             //   cout<<st<<"\n";
            }else{
                
                mx =max(mx,i-st+1);
            }
            map[s[i]]=i;
        } else{
            map[s[i]]=i;
            mx = max(mx,i-st+1);
        }
    }
    return mx;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.Arrays;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        // Guard clause for null or empty input
        if (s == null || s.isEmpty()) {
            return 0;
        }

        int maxLength = 0;
        int left = 0;

        // Tracks the last seen index of each ASCII character
        int[] lastSeenIndex = new int[128];
        Arrays.fill(lastSeenIndex, -1);

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);
            
            // If the character was seen within the current active window
            if (lastSeenIndex[currentChar] >= left) {
                left = lastSeenIndex[currentChar] + 1;
            }

            // Record/update the latest index for this character
            lastSeenIndex[currentChar] = right;

            // Calculate current valid window size
            int currentWindowLength = right - left + 1;
            maxLength = Math.max(maxLength, currentWindowLength);
        }

        return maxLength;
    }
}